potassium/potassium.py

```diff
@@ -132,69 +135,81 @@ def test_client(self):
-                    res = make_response(tb_str)
-                    res.status_code = 500
-                    res.headers['X-Endpoint-Type'] = endpoint.type
-                    return res
-
-        if endpoint.type == "background":
+            try:
+                out = endpoint.func(req)
+                res = make_response(out.json)
+                res.status_code = out.status
+                res.headers['X-Endpoint-Type'] = endpoint.type
+            except:
+                tb_str = traceback.format_exc()
+                print(colored(tb_str, "red"))
+                res = make_response(tb_str)
+                res.status_code = 500
+                res.headers['X-Endpoint-Type'] = endpoint.type
+            self._gpu_lock.release()
+        elif endpoint.type == "background":
             req = Request(
                 json=flask_request.get_json()
             )
 
             # run as threaded task
             def task(endpoint, lock, req):
-                with lock:
-                    try:
-                        endpoint.func(req)
-                    except Exception as e:
-                        # do any cleanup before re-raising user error
-                        self._write_event_chan(True)
-                        raise e
-                self._write_event_chan(True)
-
-            thread = Thread(target=task, args=(endpoint, self._lock, req))
+                try:
+                    endpoint.func(req)
+                except Exception as e:
+                    # do any cleanup before re-raising user error
+                    raise e
+                finally:
+                    self._background_task_cv.notify_all()
+                    # release lock
+                    lock.release()
+
+            thread = Thread(target=task, args=(endpoint, self._gpu_lock, req))
             thread.start()
 
             # send task start success message
             res = make_response({'started': True})
             res.headers['X-Endpoint-Type'] = endpoint.type
-            return res
+        else:
+            raise InvalidEndpointTypeException()
 
-        raise InvalidEndpointTypeException()
-
-    def _write_event_chan(self, item):
-        try:
-            self._event_chan.put(item, block=False)
-        except Full:
-            pass
+        return res
 
+    # TODO - cover depends on this being called so it should not be private
     def _read_event_chan(self) -> bool:
-        return self._event_chan.get()
+        """
+        _read_event_chan essentially waits for a background task to finish, 
+        and then returns True
+        """
+        with self._gpu_lock:
+            # wait until the background task is done
+            self._background_task_cv.wait()
+        return True
 
-    def _is_working(self):
-        return self._lock.locked()
+    # TODO - we should migrate cover to use this instead of _read_event_chan
+    def wait_for_background_task(self):
+        self._read_event_chan()
```

Comment:
the reason we chose the event_chan was for the reason cited below: with a queue you can ensure one item per release event